src/os/ui/file/addserver.js

```diff
@@ -0,0 +1,262 @@
+goog.module('os.ui.file.AddServer');
+goog.module.declareLegacyNamespace();
+
+const FileParserConfig = goog.require('os.parse.FileParserConfig');
+const Module = goog.require('os.ui.Module');
+const ImportManager = goog.require('os.ui.im.ImportManager');
+const ProviderImportLoadEventType = goog.require('os.ui.ProviderImportLoadEventType');
+const {ROOT} = goog.require('os');
+const window = goog.require('os.ui.window');
+const WindowEventType = goog.require('os.ui.WindowEventType');
+
+
+
+/**
+ * HTML ID for the Format Help windows
+ * @type {string}
+ */
+const helpWindowId = 'url-help';
+
+
+/**
+ * The addserver directive.
+ * @return {angular.Directive}
+ */
+const directive = () => {
+  return {
+    restrict: 'E',
+    replace: true,
+    scope: true,
+    templateUrl: ROOT + 'views/file/addserver.html',
+    controller: Controller,
+    controllerAs: 'ctrl'
+  };
+};
+
+
+/**
+ * The element tag for the directive.
+ * @type {string}
+ */
+const directiveTag = 'addserver';
+
+
+/**
+ * Add the directive to the module.
+ */
+Module.directive(directiveTag, [directive]);
+
+
+/**
+ * Controller for the addserver directive.
+ * @unrestricted
+ */
+class Controller {
+  /**
+   * Constructor.
+   * @param {!angular.Scope} $scope
+   * @param {!angular.JQLite} $element
+   * @ngInject
+   */
+  constructor($scope, $element) {
+    /**
+     * @type {?angular.Scope}
+     * @private
+     */
+    this.scope_ = $scope;
+
+    /**
+     * @type {?angular.JQLite}
+     * @private
+     */
+    this.element_ = $element;
+
+    /**
+     * @type {boolean}
+     */
+    this.loading = false;
+
+    /**
+     * @type {string}
+     */
+    this.serverType = '';
+
+    /**
+     * Available server type choices in the UI.
+     * @type {Array}
+     */
+    this.items = Object.values(ImportManager.getInstance().getServerTypes() || []);
+
+    $scope.$emit(WindowEventType.READY);
+    $scope.$on(ProviderImportLoadEventType.start, this.onFormLoadingStatusChange_.bind(this));
+    $scope.$on(ProviderImportLoadEventType.stop, this.onFormLoadingStatusChange_.bind(this));
+    $scope.$on('launchHelp', this.onLaunchHelp_.bind(this));
```

Comment:
Currently the help UI only shows up when adding a server via the new Add Server dialog. It would be nice if it was also displayed when adding a server via the Import Data dialog (the file open button next to Add Data). To accomplish this, I think we can do the following:

- Move `launchHelp` and `closeHelpWindow` into `ProviderImportCtrl`.
- Put `helpWindowId` on the controller. This will allow overriding classes to change the id if they want.
- We can either remove the server-specific portion of the window label (so all are consistent), or also add a `helpUiLabel` property to the controller that can be defined in overriding classes.
- Add an `$onDestroy` method to `ProviderImportCtrl` and call `closeHelpWindow` there. That will handle closing the UI when the window is closed, and when the server type is changed in Add Server (the controller will be disposed and a new one created for the different server type).
- Add a `helpUi` property to `ProviderImportCtrl` that defaults to an empty string, and can be set by overriding classes like `GeoserverImportCtrl`.
- Change `singleurlform` to only show the button when `ctrl.helpUi` is defined. This combined with the `singleurlform` reuse comment will make sure the help UI is opt-in and will display in both Add Server and Import Data for each server type that supports it.

These changes also have the benefit of moving the help window management into the same UI as the button launching help. While `$emit` and `$broadcast` are useful for passing information between directives, they are sometimes overused in ways that make the overall code more complicated/confusing.